File: Cpp/239_Sliding_Window_Maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        deque<int> dq;

        for(int i=0; i<k; ++i){
            while(!dq.empty() && nums[i]>nums[dq.back()]){
                dq.pop_back();
            }
            dq.push_back(i);
        }

        result.push_back(nums[dq.front()]);

        for(int i=k; i<nums.size(); ++i){
            if(dq.front() == i-k){
                dq.pop_front();
            }
            while(!dq.empty() && nums[i]>nums[dq.back()]){
                dq.pop_back();
            }
            dq.push_back(i);
            result.push_back(nums[dq.front()]);
        }

        return result;
    }
};

File: Kotlin/239_Sliding_Window_Maximum.kt
class Solution {
    fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
        val result = mutableListOf<Int>()
        val dq = LinkedList<Int>()
        
        for(i in 0 until k) {
            while(dq.isNotEmpty() && nums[i] > nums[dq.last()]){
                dq.pollLast()
            }
            
            dq.offerLast(i)
        }
        result.add(nums[dq.first()])
        
        for(i in k until nums.size){
            if(dq.first() == i - k){
                dq.pollFirst()
            }
            
            while(dq.isNotEmpty() && nums[i] > nums[dq.last()]){
                dq.pollLast()
            }
            
            dq.offerLast(i)
            result.add(nums[dq.first()])
        }
        
        return result.toIntArray()
    }
}